azurerm/resource_arm_monitor_diagnostic_setting.go

```diff
@@ -0,0 +1,441 @@
+										Required: true,
+									},
+
+									"days": {
+										Type:         schema.TypeInt,
+										Optional:     true,
+										ValidateFunc: validation.IntAtLeast(0),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+
+			"metric": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"category": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  true,
+						},
+
+						"retention_policy": {
+							Type:     schema.TypeList,
+							Required: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"enabled": {
+										Type:     schema.TypeBool,
+										Required: true,
+									},
+
+									"days": {
+										Type:         schema.TypeInt,
+										Optional:     true,
+										ValidateFunc: validation.IntAtLeast(0),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceArmMonitorDiagnosticSettingCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorDiagnosticSettingsClient
+	ctx := meta.(*ArmClient).StopContext
+	log.Printf("[INFO] preparing arguments for Azure ARM Diagnostic Settings.")
+
+	name := d.Get("name").(string)
+	actualResourceId := d.Get("target_resource_id").(string)
+
+	logsRaw := d.Get("log").(*schema.Set).List()
+	logs := expandMonitorDiagnosticsSettingsLogs(logsRaw)
+	metricsRaw := d.Get("metric").(*schema.Set).List()
+	metrics := expandMonitorDiagnosticsSettingsMetrics(metricsRaw)
+
+	// if no blocks are specified the API "creates" but 404's on Read
+	if len(logs) == 0 && len(metrics) == 0 {
+		return fmt.Errorf("At least one `log` or `metric` block must be specified")
+	}
+
+	properties := insights.DiagnosticSettingsResource{
+		DiagnosticSettings: &insights.DiagnosticSettings{
+			Logs:    &logs,
+			Metrics: &metrics,
+		},
+	}
+
+	valid := false
+	eventHubAuthorizationRuleId := d.Get("eventhub_authorization_rule_id").(string)
+	eventHubName := d.Get("eventhub_name").(string)
+	if eventHubAuthorizationRuleId != "" {
+		properties.DiagnosticSettings.EventHubAuthorizationRuleID = utils.String(eventHubAuthorizationRuleId)
+		properties.DiagnosticSettings.EventHubName = utils.String(eventHubName)
+		valid = true
+	}
+
+	workspaceId := d.Get("log_analytics_workspace_id").(string)
+	if workspaceId != "" {
+		properties.DiagnosticSettings.WorkspaceID = utils.String(workspaceId)
+		valid = true
+	}
+
+	storageAccountId := d.Get("storage_account_id").(string)
+	if storageAccountId != "" {
+		properties.DiagnosticSettings.StorageAccountID = utils.String(storageAccountId)
+		valid = true
+	}
+
+	if !valid {
+		return fmt.Errorf("Either a `eventhub_authorization_rule_id`, `log_analytics_workspace_id` or `storage_account_id` must be set")
```

Comment:
We could move this to a [CustomizeDiff](https://github.com/terraform-providers/terraform-provider-azurerm/blob/cadce839c18ec0d88c9152c029108e350d4892b7/azurerm/resource_arm_servicebus_namespace.go#L94)

Also, is it valid to have all of these set at the same time? or can only one be used?

Reply:
with some additional testing - there is an edge-case I missed first time around - which is that at least one setting needs to be enabled for this to provision successfully (else the API returns a 404 on creation). I've pushed a commit with logic for this